File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/ExampleArenaCreator.java
package com.gmail.borlandlp.minigamesdtools.arena;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.arena.chunkloader.ChunkLoaderController;
import com.gmail.borlandlp.minigamesdtools.arena.chunkloader.ChunksLoader;
import com.gmail.borlandlp.minigamesdtools.arena.commands.ArenaCommandWatcher;
import com.gmail.borlandlp.minigamesdtools.arena.commands.CommandWatcherCreatorHub;
import com.gmail.borlandlp.minigamesdtools.conditions.AbstractCondition;
import com.gmail.borlandlp.minigamesdtools.conditions.ConditionsChain;
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath;
import com.gmail.borlandlp.minigamesdtools.creator.*;
import com.gmail.borlandlp.minigamesdtools.arena.gui.hotbar.HotbarController;
import com.gmail.borlandlp.minigamesdtools.arena.gui.providers.GUIController;
import com.gmail.borlandlp.minigamesdtools.arena.gui.providers.GUIProvider;
import com.gmail.borlandlp.minigamesdtools.arena.scenario.ScenarioChainController;
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamController;
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamProvider;
import com.gmail.borlandlp.minigamesdtools.util.ArenaUtils;
import com.gmail.borlandlp.minigamesdtools.util.Glowing;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.configuration.ConfigurationSection;

import java.util.*;

@CreatorInfo(creatorId = "default_arena")
public class ExampleArenaCreator implements Creator {
    @Override
    public List<String> getDataProviderRequiredFields() {
        return new ArrayList<>();
    }

    @Override
    public ArenaBase create(String ID, AbstractDataProvider dataProvider) throws Exception {
        String debugPrefix = "[" + ID + "] ";
        ConfigurationSection arenaConfig = MinigamesDTools.Companion.getInstance().getConfigProvider().getEntity(ConfigPath.ARENA_FOLDER, ID).getData();
        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " started loading...");

        ArenaBase.Builder builder = ArenaBase.newBuilder();

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " load chunkLoaders...");
        ChunkLoaderController chunkLoaderController = new ChunkLoaderController();
        chunkLoaderController.setArena(builder.getArena());
        Map<String, ChunksLoader> chunksLoaderMap = new HashMap<>();
        if(arenaConfig.contains("chunk_loaders")) {
            for (String loaderName : arenaConfig.getConfigurationSection("chunk_loaders").getKeys(false)) {
                if(arenaConfig.contains("chunk_loaders." + loaderName + ".min_point_xz") && arenaConfig.contains("chunk_loaders." + loaderName + ".max_point_xz")) {
                    String[] minXZ = arenaConfig.get("chunk_loaders." + loaderName + ".min_point_xz").toString().split(":");
                    String[] maxXZ = arenaConfig.get("chunk_loaders." + loaderName + ".max_point_xz").toString().split(":");
                    World world = Bukkit.getWorld(arenaConfig.get("chunk_loaders." + loaderName + ".world").toString());
                    Location min_loc = world.getBlockAt(Integer.parseInt(minXZ[0]), 1, Integer.parseInt(minXZ[1])).getLocation();
                    Location max_loc = world.getBlockAt(Integer.parseInt(maxXZ[0]), 1, Integer.parseInt(maxXZ[1])).getLocation();

                    chunksLoaderMap.put(loaderName, MinigamesDTools.Companion.getInstance().getChunkLoaderCreator().buildChunkLoader(world, min_loc, max_loc));
                } else {
                    Debug.print(Debug.LEVEL.NOTICE, "ChunkLoaderConfig does not contain a field 'min_point_xz' or 'max_point_xz' -> Skipping it.");
                }
            }

            chunkLoaderController.setLoaders(chunksLoaderMap);
        } else {
            Debug.print(Debug.LEVEL.NOTICE, "ArenaConfig does not contain a field 'ChunkLoaders' -> Skipping it.");
        }
        builder.setChunkLoaderController(chunkLoaderController);

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " load GUIController...");
        GUIController guiController = new GUIController();
        guiController.setArena(builder.getArena());
        for (String GUI_ID : arenaConfig.getStringList("gui_provider")) {
            GUIProvider guiProvider = MinigamesDTools.Companion.getInstance().getGuiCreatorHub().createGuiProvider(GUI_ID, new DataProvider());
            guiProvider.setArena(builder.getArena());
            guiController.addProvider(guiProvider);
        }
        builder.setGuiController(guiController);

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " load ScenarioChain...");
        AbstractDataProvider abstractDataProvider = new DataProvider();
        abstractDataProvider.set("arena_instance", builder.getArena());
        ScenarioChainController scenarioChainController = MinigamesDTools.Companion.getInstance().getScenarioChainCreatorHub().createChain(arenaConfig.get("scenarios_chain").toString(), abstractDataProvider);
        builder.setScenarioChainController(scenarioChainController);

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " load teams...");
        TeamController teamController = new TeamController(builder.getArena());
        for(String teamID : arenaConfig.getStringList("teams")) {
            TeamProvider teamProvider = MinigamesDTools.Companion.getInstance().getTeamCreatorHub().createTeam(teamID, new DataProvider());
            teamProvider.setArena(builder.getArena());
            teamController.addTeam(teamProvider);
        }
        builder.setTeamController(teamController);

        //conditions
        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " load conditions...");
        List<AbstractCondition> conditions = new ArrayList<>();
        for (String conditionId : arenaConfig.getStringList("join_conditions")) {
            conditions.add(MinigamesDTools.Companion.getInstance().getConditionsCreatorHub().createCondition(conditionId, new DataProvider()));
        }
        builder.setJoinConditionsChain(new ConditionsChain(conditions));

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " load hotbar...");
        HotbarController hotbarController = new HotbarController();
        hotbarController.setArena(builder.getArena());
        if(Boolean.parseBoolean(arenaConfig.get("interactive_hotbar.enabled").toString())) {
            String hotbarID = arenaConfig.get("interactive_hotbar.hotbar_id").toString();
            hotbarController.setDefaultHotbarId(hotbarID);
            hotbarController.setEnabled(true);
        } else {
            Debug.print(Debug.LEVEL.NOTICE, "Hotbar is disabled for Arena " + ID);
        }
        builder.setHotbarController(hotbarController);

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + "generate random ID...");
        builder.setGameId(ArenaUtils.generateRandID().substring(0, 14));

        builder.setName(ID);

        // gen colors
        /*
         * GlowAPI contains the last element color "None". We exclude it.
         * */
        if( builder.getArena().getTeamController().getTeams().size() > Glowing.getAvailableColors().size()-1) {
            throw new Exception("Max team limit reached! [limit=" + (Glowing.getAvailableColors().size()-1) + "]");
        }
        List<String> usedColors = new ArrayList<>();
        for (TeamProvider team : builder.getArena().getTeamController().getTeams()) {
            ChatColor color;
            String strColor;
            do {
                int rnd = new Random().nextInt(Glowing.getAvailableColors().size() - 2); // exclude last color GlowAPI.Color -> "None"
                strColor = Glowing.getAvailableColors().get(rnd);
            } while(usedColors.contains(strColor));
            usedColors.add(strColor);
            team.setColor(strColor);
        }

        // <GameRules>
        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + " load options...");
        GameRules gameRules = new GameRules();
        gameRules.roundTime = Integer.parseInt(arenaConfig.get("roundTime").toString());
        gameRules.maxRounds = Integer.parseInt(arenaConfig.get("maxRounds").toString());
        gameRules.playerCanItemDrop = Boolean.parseBoolean(arenaConfig.get("playerItemDrop").toString());
        gameRules.playerCanItemPickup = Boolean.parseBoolean(arenaConfig.get("playerItemPickUp").toString());
        gameRules.beforeRoundStartingWaitDuration = Integer.parseInt(arenaConfig.get("countdown.beforeFight.duration").toString());
        gameRules.beforeArenaTeleportWaitDuration = Integer.parseInt(arenaConfig.get("countdown.beforeTeleport.duration").toString());
        gameRules.beforeFightDisableMoving = Boolean.parseBoolean(arenaConfig.get("countdown.beforeFight.disableMoving").toString());
        gameRules.hungerDisable = Boolean.parseBoolean(arenaConfig.get("playerHungerDisable").toString());
        gameRules.minPlayersToStart = Integer.parseInt(arenaConfig.get("min_players_to_start").toString());
        if(arenaConfig.contains("regain_health")) {
            gameRules.playerCanRegainHealth = arenaConfig.get("regain_health").toString().equalsIgnoreCase("true");
        }
        builder.setGameRules(gameRules);
        // </GameRules>

        ArenaBase arenaTemplate = builder.build();
        arenaTemplate.setEnabled(arenaConfig.get("enabled").toString().equalsIgnoreCase("true"));

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + "register phase components...");
        // register
        arenaTemplate.getPhaseComponentController().register(arenaTemplate.getChunkLoaderController());
        arenaTemplate.getPhaseComponentController().register(arenaTemplate.getTeamController());
        arenaTemplate.getPhaseComponentController().register(arenaTemplate.getScenarioChainController());
        arenaTemplate.getPhaseComponentController().register(arenaTemplate.getGuiController());
        arenaTemplate.getPhaseComponentController().register(arenaTemplate.getHotbarController());
        arenaTemplate.getPhaseComponentController().register(arenaTemplate.getHandlersController());

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + "try init command watcher...");
        // ArenaCommandWatcher
        if(arenaConfig.contains("commands")) {
            AbstractDataProvider cDataProvider = new DataProvider();
            cDataProvider.set("arena_instance", arenaTemplate);

            List<String[]> blacklisted = new ArrayList<>();
            arenaConfig.getStringList("commands.blacklist").forEach(e -> blacklisted.add(e.split(" ")));
            cDataProvider.set("blacklist_rules", blacklisted);

            List<String[]> whitelisted = new ArrayList<>();
            arenaConfig.getStringList("commands.whitelist").forEach(e -> whitelisted.add(e.split(" ")));
            cDataProvider.set("whitelist_rules", whitelisted);

            String handlerID = arenaConfig.get("commands.handler").toString();
            CommandWatcherCreatorHub creatorHub = MinigamesDTools.Companion.getInstance().getCommandWatcherCreatorHub();
            ArenaCommandWatcher watcher = creatorHub.createCommandWatcher(handlerID, cDataProvider);

            arenaTemplate.getPhaseComponentController().register(watcher);
        }

        Debug.print(Debug.LEVEL.NOTICE, debugPrefix + "loading is done!");

        return arenaTemplate;
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/ExampleArenaCreator.kt
package com.gmail.borlandlp.minigamesdtools.arena

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools.Companion.instance
import com.gmail.borlandlp.minigamesdtools.arena.chunkloader.ChunkLoaderController
import com.gmail.borlandlp.minigamesdtools.arena.chunkloader.ChunksLoader
import com.gmail.borlandlp.minigamesdtools.arena.gui.hotbar.HotbarController
import com.gmail.borlandlp.minigamesdtools.arena.gui.providers.GUIController
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamController
import com.gmail.borlandlp.minigamesdtools.conditions.AbstractCondition
import com.gmail.borlandlp.minigamesdtools.conditions.ConditionsChain
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath
import com.gmail.borlandlp.minigamesdtools.creator.AbstractDataProvider
import com.gmail.borlandlp.minigamesdtools.creator.Creator
import com.gmail.borlandlp.minigamesdtools.creator.CreatorInfo
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider
import com.gmail.borlandlp.minigamesdtools.util.ArenaUtils.generateRandID
import com.gmail.borlandlp.minigamesdtools.util.Glowing.availableColors
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import java.util.*
import java.util.function.Consumer

@CreatorInfo(creatorId = "default_arena")
class ExampleArenaCreator : Creator() {
    @Throws(Exception::class)
    override fun create(id: String, dataProvider: AbstractDataProvider): ArenaBase {
        val debugPrefix = "[$id] "
        val arenaConfig =
            instance!!.configProvider!!.getEntity(ConfigPath.ARENA_FOLDER, id)!!.data
        print(
            Debug.LEVEL.NOTICE,
            "$debugPrefix started loading..."
        )
        val builder = ArenaBase.newBuilder()
        print(
            Debug.LEVEL.NOTICE,
            "$debugPrefix load chunkLoaders..."
        )
        val chunkLoaderController = ChunkLoaderController()
        chunkLoaderController.arena = builder.arena
        val chunksLoaderMap: MutableMap<String, ChunksLoader> =
            HashMap()
        if (arenaConfig.contains("chunk_loaders")) {
            for (loaderName in arenaConfig.getConfigurationSection("chunk_loaders").getKeys(false)) {
                if (arenaConfig.contains("chunk_loaders.$loaderName.min_point_xz") && arenaConfig.contains("chunk_loaders.$loaderName.max_point_xz")) {
                    val minXZ =
                        arenaConfig["chunk_loaders.$loaderName.min_point_xz"].toString().split(":")
                            .toTypedArray()
                    val maxXZ =
                        arenaConfig["chunk_loaders.$loaderName.max_point_xz"].toString().split(":")
                            .toTypedArray()
                    val world =
                        Bukkit.getWorld(arenaConfig["chunk_loaders.$loaderName.world"].toString())
                    val min_loc =
                        world.getBlockAt(minXZ[0].toInt(), 1, minXZ[1].toInt()).location
                    val max_loc =
                        world.getBlockAt(maxXZ[0].toInt(), 1, maxXZ[1].toInt()).location
                    chunksLoaderMap[loaderName] =
                        instance!!.chunkLoaderCreator!!.buildChunkLoader(world, min_loc, max_loc)
                } else {
                    print(
                        Debug.LEVEL.NOTICE,
                        "ChunkLoaderConfig does not contain a field 'min_point_xz' or 'max_point_xz' -> Skipping it."
                    )
                }
            }
            chunkLoaderController.loaders = chunksLoaderMap
        } else {
            print(
                Debug.LEVEL.NOTICE,
                "ArenaConfig does not contain a field 'ChunkLoaders' -> Skipping it."
            )
        }
        builder.setChunkLoaderController(chunkLoaderController)
        print(
            Debug.LEVEL.NOTICE,
            "$debugPrefix load GUIController..."
        )
        val guiController = GUIController()
        guiController.arena = builder.arena
        for (GUI_ID in arenaConfig.getStringList("gui_provider")) {
            val guiProvider =
                instance!!.guiCreatorHub!!.createGuiProvider(GUI_ID!!, DataProvider())
            guiProvider.arena = builder.arena
            guiController.addProvider(guiProvider)
        }
        builder.setGuiController(guiController)
        print(
            Debug.LEVEL.NOTICE,
            "$debugPrefix load ScenarioChain..."
        )
        val abstractDataProvider: AbstractDataProvider = DataProvider()
        abstractDataProvider["arena_instance"] = builder.arena
        val scenarioChainController =
            instance!!.scenarioChainCreatorHub!!.createChain(
                arenaConfig["scenarios_chain"].toString(),
                abstractDataProvider
            )
        builder.setScenarioChainController(scenarioChainController)
        print(
            Debug.LEVEL.NOTICE,
            "$debugPrefix load teams..."
        )
        val teamController = TeamController(builder.arena)
        for (teamID in arenaConfig.getStringList("teams")) {
            val teamProvider =
                instance!!.teamCreatorHub!!.createTeam(teamID!!, abstractDataProvider)
            teamProvider.arena = builder.arena
            teamController.addTeam(teamProvider)
        }
        builder.setTeamController(teamController)
        //conditions
        print(
            Debug.LEVEL.NOTICE,
            "$debugPrefix load conditions..."
        )
        val conditions: MutableList<AbstractCondition> = ArrayList()
        for (conditionId in arenaConfig.getStringList("join_conditions")) {
            conditions.add(
                instance!!.conditionsCreatorHub!!.createCondition(
                    conditionId!!,
                    abstractDataProvider
                )
            )
        }
        builder.setJoinConditionsChain(ConditionsChain(conditions))
        print(
            Debug.LEVEL.NOTICE,
            "$debugPrefix load hotbar..."
        )
        val hotbarController = HotbarController()
        hotbarController.arena = builder.arena
        if (arenaConfig["interactive_hotbar.enabled"].toString().toBoolean()) {
            val hotbarID = arenaConfig["interactive_hotbar.hotbar_id"].toString()
            hotbarController.defaultHotbarId = hotbarID
            hotbarController.isEnabled = true
        } else {
            print(
                Debug.LEVEL.NOTICE,
                "Hotbar is disabled for Arena $id"
            )
        }
        builder.setHotbarController(hotbarController)
        print(
            Debug.LEVEL.NOTICE,
            debugPrefix + "generate random ID..."
        )
        builder.setGameId(generateRandID().substring(0, 14))
        builder.setName(id)
        // gen colors
        /*
         * GlowAPI contains the last element color "None". We exclude it.
         * */
        if (builder.arena.teamController!!.teams.size > availableColors.size - 1) {
            throw Exception("Max team limit reached! [limit=" + (availableColors.size - 1) + "]")
        }
        val usedColors: MutableList<String> = ArrayList()
        for (team in builder.arena.teamController!!.teams) {
            var color: ChatColor
            var strColor: String
            do {
                val rnd = Random()
                    .nextInt(availableColors.size - 2) // exclude last color GlowAPI.Color -> "None"
                strColor = availableColors[rnd]
            } while (usedColors.contains(strColor))
            usedColors.add(strColor)
            team.color = strColor
        }
        // <GameRules>
        print(Debug.LEVEL.NOTICE, "$debugPrefix load options...")
        val gameRules = GameRules()
        gameRules.roundTime = arenaConfig["roundTime"].toString().toInt()
        gameRules.maxRounds = arenaConfig["maxRounds"].toString().toInt()
        gameRules.playerCanItemDrop = arenaConfig["playerItemDrop"].toString().toBoolean()
        gameRules.playerCanItemPickup = arenaConfig["playerItemPickUp"].toString().toBoolean()
        gameRules.beforeRoundStartingWaitDuration = arenaConfig["countdown.beforeFight.duration"].toString().toInt()
        gameRules.beforeArenaTeleportWaitDuration =
            arenaConfig["countdown.beforeTeleport.duration"].toString().toInt()
        gameRules.beforeFightDisableMoving = arenaConfig["countdown.beforeFight.disableMoving"].toString().toBoolean()
        gameRules.hungerDisable = arenaConfig["playerHungerDisable"].toString().toBoolean()
        gameRules.minPlayersToStart = arenaConfig["min_players_to_start"].toString().toInt()
        if (arenaConfig.contains("regain_health")) {
            gameRules.playerCanRegainHealth =
                arenaConfig["regain_health"].toString().equals("true", ignoreCase = true)
        }
        builder.setGameRules(gameRules)
        // </GameRules>
        val arenaTemplate = builder.build()
        arenaTemplate.isEnabled = arenaConfig["enabled"].toString().equals("true", ignoreCase = true)
        print(
            Debug.LEVEL.NOTICE,
            debugPrefix + "register phase components..."
        )
        // register
        arenaTemplate.phaseComponentController.register(arenaTemplate.chunkLoaderController as ArenaPhaseComponent)
        arenaTemplate.phaseComponentController.register(arenaTemplate.teamController as ArenaPhaseComponent)
        arenaTemplate.phaseComponentController.register(arenaTemplate.scenarioChainController as ArenaPhaseComponent)
        arenaTemplate.phaseComponentController.register(arenaTemplate.guiController as ArenaPhaseComponent)
        arenaTemplate.phaseComponentController.register(arenaTemplate.hotbarController as ArenaPhaseComponent)
        arenaTemplate.phaseComponentController.register(arenaTemplate.handlersController)
        print(
            Debug.LEVEL.NOTICE,
            debugPrefix + "try init command watcher..."
        )
        // ArenaCommandWatcher
        if (arenaConfig.contains("commands")) {
            val cDataProvider: AbstractDataProvider = DataProvider()
            cDataProvider["arena_instance"] = arenaTemplate
            val blacklisted: MutableList<Array<String>> =
                ArrayList()
            arenaConfig.getStringList("commands.blacklist")
                .forEach(Consumer { e: String -> blacklisted.add(e.split(" ").toTypedArray()) })
            cDataProvider["blacklist_rules"] = blacklisted
            val whitelisted: MutableList<Array<String>> =
                ArrayList()
            arenaConfig.getStringList("commands.whitelist")
                .forEach(Consumer { e: String -> whitelisted.add(e.split(" ").toTypedArray()) })
            cDataProvider["whitelist_rules"] = whitelisted
            val handlerID = arenaConfig["commands.handler"].toString()
            val creatorHub = instance!!.commandWatcherCreatorHub
            val watcher = creatorHub!!.createCommandWatcher(handlerID, cDataProvider)
            arenaTemplate.phaseComponentController.register(watcher)
        }
        print(
            Debug.LEVEL.NOTICE,
            debugPrefix + "loading is done!"
        )
        return arenaTemplate
    }
}